drivers/pinctrl/pinctrl_esp32.c

```diff
@@ -0,0 +1,286 @@
+		}
+
+		gpio_ll_output_enable(&GPIO, io_pin);
+		esp_rom_gpio_matrix_out(io_pin, SIG_GPIO_OUT_IDX, false, false);
+	} else {
+		gpio_ll_output_disable(&GPIO, io_pin);
+	}
+
+	if (flags & ESP32_DIR_INP_FLAG) {
+		gpio_ll_input_enable(&GPIO, io_pin);
+	} else {
+		gpio_ll_input_disable(&GPIO, io_pin);
+	}
+
+end:
+	return ret;
+}
+
+static int esp32_pin_configure(const uint32_t pin_mux, const uint32_t pin_cfg)
+{
+	uint32_t port_addr;
+	uint32_t pin_num = ESP32_PIN_NUM(pin_mux);
+	uint32_t sig_in = ESP32_PIN_SIGI(pin_mux);
+	uint32_t sig_out = ESP32_PIN_SIGO(pin_mux);
+	uint32_t flags = 0;
+
+	if (ESP32_PORT_IDX(pin_num) >= esp32_gpio_ports_cnt) {
+		return -EINVAL;
+	}
+
+	port_addr = esp32_gpio_ports_addrs[ESP32_PORT_IDX(pin_num)];
+
+	if (port_addr == ESP32_INVALID_PORT_ADDR) {
+		return -EINVAL;
+	}
+
+	switch (ESP32_PIN_BIAS(pin_cfg)) {
+	case ESP32_PULL_UP:
+		flags |= ESP32_PULL_UP_FLAG;
+		break;
+	case ESP32_PULL_DOWN:
+		flags |= ESP32_PULL_DOWN_FLAG;
+		break;
+	default:
+		break;
+	}
+
+	switch (ESP32_PIN_DRV(pin_cfg)) {
+	case ESP32_PUSH_PULL:
+		flags |= ESP32_PUSH_PULL_FLAG;
+		break;
+	case ESP32_OPEN_DRAIN:
+		flags |= ESP32_OPEN_DRAIN_FLAG;
+		break;
+	default:
+		break;
+	}
+
+	if (sig_in == ESP_SIG_INVAL && sig_out == ESP_SIG_INVAL) {
+		return -ENOTSUP;
+	}
+
+	if (sig_in != ESP_SIG_INVAL) {
+		flags |= ESP32_DIR_INP_FLAG;
+	}
+
+	if (sig_out != ESP_SIG_INVAL) {
+		flags |= ESP32_DIR_OUT_FLAG;
+	}
+
+	switch (ESP32_PIN_MODE_OUT(pin_cfg)) {
+	case ESP32_PIN_OUT_HIGH:
+		flags |= ESP32_PIN_OUT_HIGH_FLAG;
+		break;
+	case ESP32_PIN_OUT_LOW:
+		flags |= ESP32_PIN_OUT_LOW_FLAG;
+		break;
+	default:
+		break;
+	}
+
+	if (flags & ESP32_PIN_OUT_HIGH_FLAG) {
+		if (ESP32_PORT_IDX(pin_num) == 0) {
+			gpio_dev_t *const gpio_dev =
+				(gpio_dev_t *)DT_REG_ADDR(DT_NODELABEL(gpio0));
+			gpio_dev->out_w1ts = pin_num;
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio1), okay)
+		} else {
+			gpio_dev_t *const gpio_dev =
+				(gpio_dev_t *)DT_REG_ADDR(DT_NODELABEL(gpio1));
+			gpio_dev->out1_w1ts.data = pin_num;
+#endif
+		}
+	}
+
+	if (flags & ESP32_PIN_OUT_LOW_FLAG) {
+		if (ESP32_PORT_IDX(pin_num) == 0) {
+			gpio_dev_t *const gpio_dev =
+				(gpio_dev_t *)DT_REG_ADDR(DT_NODELABEL(gpio0));
+			gpio_dev->out_w1tc = pin_num;
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio1), okay)
+		} else {
+			gpio_dev_t *const gpio_dev =
+				(gpio_dev_t *)DT_REG_ADDR(DT_NODELABEL(gpio1));
+			gpio_dev->out1_w1tc.data = pin_num;
+#endif
+		}
+	}
+
+	esp32_pin_apply_config(pin_num, flags);
+
+	if (flags & ESP32_DIR_OUT_FLAG) {
+		esp_rom_gpio_matrix_out(pin_num, sig_out, 0, 0);
+	}
+
+	if (flags & ESP32_DIR_INP_FLAG) {
+		esp_rom_gpio_matrix_in(pin_num, sig_in, 0);
+	}
+
+	return 0;
+}
+
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
+			   uintptr_t reg)
+{
+	uint32_t pin_mux, pin_cfg;
+	int ret = 0;
+
+	ARG_UNUSED(reg);
+
+	for (uint8_t i = 0U; i < pin_cnt; i++) {
+		pin_mux = pins[i].pinmux;
```

Comment:
```suggestion
	ARG_UNUSED(reg);

	for (uint8_t i = 0U; i < pin_cnt; i++) {
		int ret;

		pin_mux = pins[i].pinmux;
```

Reply:
Though, in general, it is good practice to keep variables restricted to the scope they are being used. Here, I'd rather keep it out to avoid variable redefinitions or stack unwinding in different compiling optimization scenarios (including, eventually, debugging).